### PR TITLE
Fix macOS build 

### PR DESCRIPTION
**Pull request by Lucky-Chang**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
fix build on macOS

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by Lucky-Chang** [comment]
https://github.com/ClickHouse/ClickHouse/issues/40937#issue-1360096156

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -7217,7 +7217,7 @@ bool StorageReplicatedMergeTree::addOpsToDropAllPartsInPartition(
 }
 
 void StorageReplicatedMergeTree::dropAllPartsInPartitions(
-    zkutil::ZooKeeper & zookeeper, const Strings partition_ids, std::vector<LogEntryPtr> & entries, ContextPtr query_context, bool detach)
+    zkutil::ZooKeeper & zookeeper, const Strings & partition_ids, std::vector<LogEntryPtr> & entries, ContextPtr query_context, bool detach)
```
Comment (serxa) [inline src/Storages/StorageReplicatedMergeTree.cpp:7220]: nice catch

**Review by serxa — APPROVED** [review]
LGTM, let's look if CI agree with all typo fixes in logs and error messages

File: utils/self-extracting-executable/post_build.sh [hunk 2]
```diff
@@ -1,6 +1,10 @@
 padding="               "
 if [[ $OSTYPE == 'darwin'* ]]; then
-    sz="$(stat -f %z 'decompressor')"
+    if which gstat; then
```
Comment (rschu1ze) [inline utils/self-extracting-executable/post_build.sh:3]: The GNU and the Mac version stat support different command line arguments (`-c %s` vs `-f %z`). On Mac, we are now checking for `gstat` which is the GNU version which can be installed via brew. Since installing `gstat` via brew does not remove or disable the build-in Mac version of `stat`, I wonder how this PR "fixes the macOS build" (commit message)?

File: src/Loggers/CMakeLists.txt [hunk 3]
```diff
@@ -2,9 +2,9 @@ add_headers_and_sources(loggers .)
 
 # Standard version depends on DBMS and works with text log
 add_library(loggers ${loggers_sources} ${loggers_headers})
-target_compile_definitions(loggers PUBLIC WITH_TEXT_LOG=1)
 target_link_libraries(loggers PRIVATE dbms clickhouse_common_io)
 
 # Lightweight version doesn't work with textlog and also doesn't depend on DBMS
 add_library(loggers_no_text_log ${loggers_sources} ${loggers_headers})
+target_compile_definitions(loggers_no_text_log PUBLIC WITHOUT_TEXT_LOG=1)
```
Comment (novikd) [inline src/Loggers/CMakeLists.txt:9]: @rschu1ze AFAIU this line is supposed to fix the build, but to be honest I don't understand why. For me it seems to have absolutely the same behaviour as before because `#ifdef WITH_TEXT_LOG` was replaced with `ifndef WITHOUT_TEXT_LOG`.

File: utils/self-extracting-executable/post_build.sh (hunk 2 above)
Comment (Lucky-Chang) [reply utils/self-extracting-executable/post_build.sh:3]: as you say, i installed gnu coreutils by homebrew in my macOS. and `gstat` is the default alternative for `stat`

File: src/Loggers/CMakeLists.txt (hunk 3 above)
Comment (Lucky-Chang) [reply src/Loggers/CMakeLists.txt:9]: i figure that the main binary include 'Loggers.h' in different way, in most case, the include path is implicated by loggers library. But some directly include the 'Loggers.h'. so the `WITH_TEXT_LOG` definition differ in the main binary.
I just reverse `WITH_TEXT_LOG` to `WITHOUT_TEXT_LOG`. so in the main library all 'Loggers.h'  has no `WITHOUT_TEXT_LOG`. it works out.

**Comment by Lucky-Chang** [comment]
exactly the same problem with https://github.com/ClickHouse/ClickHouse/issues/40398

File: utils/self-extracting-executable/post_build.sh (hunk 2 above)
Comment (rschu1ze) [reply utils/self-extracting-executable/post_build.sh:3]: Instead of making the script more complex, we could continue to call `stat` even if the user installed `gstat`.

File: src/Loggers/CMakeLists.txt (hunk 3 above)
Comment (rschu1ze) [reply src/Loggers/CMakeLists.txt:9]: The fix is a bit weird but generally okay, IMHO.

It would be great to understand the problem in more detail.

@Lucky-Chang Could you please give the detailed steps for reproduction: What is the exact command line for cmake? Please try to use homebrew's LLVM/Clang/CMake as described in the [documentation](https://clickhouse.com/docs/en/development/build-osx) instead of Apple's XCode LLVM/Clang (as used in #40398).

> i figure that the main binary include 'Loggers.h' in different way, in most case, the include path is implicated by loggers library. But some directly include the 'Loggers.h'. so the WITH_TEXT_LOG definition differ in the main binary.

- By "main binary", do you mean programs/server/Server.cpp?
- What do you mean "the loggers library implicates the include path"? More specifically, which files get included exactly and where exactly does that happen?
- Could you give an example for the files which "directly include Loggers.h"?

I think I have a clue about the cause of the problem. Sorry for asking all these questions but depending on the details, there could be potentially a better (more consistent) fix.

Comment (novikd) [reply src/Loggers/CMakeLists.txt:9]: It took me quite some time to understand how it fix the code. The simple answer `Loggers` class definition violates ODR (introduced in #35031).
Explanaition:
1. `Loggers` class is build in 2 binaries: `loggers` and `loggers_no_text_log`.
2. `loggers_no_text_log` is used only for clickhouse-keeper.
3. Everything else is linked against `loggers`, but compiled without `WITH_TEXT_LOG` macro.

For example, in BaseDeamon.h we have `#include <Loggers/Loggers.h>`. BaseDeamon.cpp is never compiled with `WITH_TEXT_LOG` macro defined, but linked with `loggers`.

File: src/Loggers/Loggers.h [hunk 4]
```diff
@@ -7,7 +7,7 @@
 #include <Poco/Util/Application.h>
 #include "OwnSplitChannel.h"
 
-#ifdef WITH_TEXT_LOG
+#ifndef WITHOUT_TEXT_LOG
```
Comment (novikd) [inline src/Loggers/Loggers.h:10]: This now leads `Loggers` to be expected with text log in clickhouse-keeper, but linked with `loggers_no_text_log`. The problem is not fixed.

Please, make a separate PR with typo fixes, because it's hard to read this PR and to understand what is the actual fix.

**Review by novikd — CHANGES_REQUESTED** [review]

File: utils/self-extracting-executable/post_build.sh (hunk 2 above)
Comment (Lucky-Chang) [reply utils/self-extracting-executable/post_build.sh:3]: the problem is directly call stat will fail, because it's actually 'gstat', the running params are different

File: src/Loggers/Loggers.h (hunk 4 above)
Comment (Lucky-Chang) [reply src/Loggers/Loggers.h:10]: src/CMakeList.txt 319~320
dbms_target_include_directories (PUBLIC "${ClickHouse_SOURCE_DIR}/src" "${ClickHouse_BINARY_DIR}/src")
target_include_directories (clickhouse_common_io PUBLIC "${ClickHouse_SOURCE_DIR}/src" "${ClickHouse_BINARY_DIR}/src")

this two lines expose all .h （include Loggers.h）to  all build libraries (common-io and all dmbs)。during all these libraries building they just not see 'WITH_TEXT_LOG' compile definition。HOW we could make sure some library will 
inherit the compile definition from the loggers (or loggers_no_text_log) library? it must be explicit link loggers (modern cmake target link rules), in the project there only two libraries do this:
target_link_libraries (daemon PUBLIC loggers)
target_link_libraries(clickhouse-keeper PRIVATE loggers_no_text_log)

At last clickhouse-server links daemon library (with `WITH_TEXT_LOG`) and other comon_io / dbms libraries (without `WITH_TEXT_LOG`), the Loggers class definition conflicts.

Above is my explanation. as for the mini-clickhouse-keeper without text log, its link dependencies have no change,  will keep small. But may still have problem, becase clickhouse-keeper links loggers_no_text_log and clickhouse_common_io, the compile definition (`WITH_TEXT_LOG`) in loggers_no_text_log not exposed to clickhouse_common_io.
target_link_libraries(clickhouse-keeper
        PRIVATE
           ...
            loggers_no_text_log
            clickhouse_common_io
    )

At last，it seems no easy way to fix this problem gracefully  now, just the whole compile target dependency is not struct in modern CMake way https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1

Comment (Lucky-Chang) [reply src/Loggers/Loggers.h:10]: i make a separate PR for typo fix here, https://github.com/ClickHouse/ClickHouse/pull/40984

**Comment by novikd** [comment]
Problem is fixed in #41060
